cardano-crypto-class/src/Cardano/Crypto/KES/Sum.hs

```diff
@@ -0,0 +1,337 @@
+  , SigKES (..)
+
+    -- * Type aliases for powers of binary sums
+  , Sum0KES
+  , Sum1KES
+  , Sum2KES
+  , Sum3KES
+  , Sum4KES
+  , Sum5KES
+  , Sum6KES
+  , Sum7KES
+  ) where
+
+import           Data.Word (Word8)
+import           Data.Proxy (Proxy(..))
+import           Data.Typeable (Typeable)
+import           GHC.Generics (Generic)
+import qualified Data.ByteString as BS
+import           Control.Monad (guard)
+
+import           Cardano.Prelude (NoUnexpectedThunks)
+import           Cardano.Binary (FromCBOR (..), ToCBOR (..))
+
+import           Cardano.Crypto.Seed
+import           Cardano.Crypto.Hash.Class
+import           Cardano.Crypto.KES.Class
+import           Cardano.Crypto.KES.Single (SingleKES)
+
+
+-- | A 2^0 period KES
+type Sum0KES d   = SingleKES d
+
+-- | A 2^1 period KES
+type Sum1KES d h = SumKES h (Sum0KES d)
+
+-- | A 2^2 period KES
+type Sum2KES d h = SumKES h (Sum1KES d h)
+
+-- | A 2^3 period KES
+type Sum3KES d h = SumKES h (Sum2KES d h)
+
+-- | A 2^4 period KES
+type Sum4KES d h = SumKES h (Sum3KES d h)
+
+-- | A 2^5 period KES
+type Sum5KES d h = SumKES h (Sum4KES d h)
+
+-- | A 2^6 period KES
+type Sum6KES d h = SumKES h (Sum5KES d h)
+
+-- | A 2^7 period KES
+type Sum7KES d h = SumKES h (Sum6KES d h)
+
+
+-- | A composition of two KES schemes to give a KES scheme with the sum of
+-- the time periods.
+--
+-- While we could do this with two independent KES schemes (i.e. two types)
+-- we only need it for two instances of the same scheme, and we save
+-- substantially on the size of the type and runtime dictionaries if we do it
+-- this way, especially when we start applying it recursively.
+--
+data SumKES h d
+
+instance (KESAlgorithm d, HashAlgorithm h, Typeable d)
+      => KESAlgorithm (SumKES h d) where
+
+
+    --
+    -- Key and signature types
+    --
+
+    -- | From Section 3,1:
+    --
+    -- The verification key @vk@ for the sum scheme is the hash of the
+    -- verification keys @vk_0, vk_1@ of the two constituent schemes.
+    --
+    newtype VerKeyKES (SumKES h d) =
+              VerKeySumKES (Hash h (VerKeyKES d, VerKeyKES d))
+        deriving Generic
+
+    -- | From Figure 3: @(sk_0, r_1, vk_0, vk_1)@
+    --
+    data SignKeyKES (SumKES h d) =
+           SignKeySumKES !(SignKeyKES d)
+                         !Seed
+                         !(VerKeyKES d)
+                         !(VerKeyKES d)
+        deriving Generic
+
+    -- | From Figure 3: @(sigma, vk_0, vk_1)@
+    --
+    data SigKES (SumKES h d) =
+           SigSumKES !(SigKES d)
+                     !(VerKeyKES d)
+                     !(VerKeyKES d)
+        deriving Generic
+
+
+    --
+    -- Metadata and basic key operations
+    --
+
+    algorithmNameKES _ = algorithmNameKES (Proxy :: Proxy d) ++ "x2"
+
+    deriveVerKeyKES (SignKeySumKES _ _ vk_0 vk_1) =
+        VerKeySumKES (hashPairOfVKeys (vk_0, vk_1))
+
+    -- The verification key in this scheme is actually a hash already
+    -- But the hashVerKeyKES 
```

Comment:
:grin: :grin: :grin:

Reply:
I swear I had a thought :smiley: 

Now fixed

> The verification key in this scheme is actually a hash already however the type of hashVerKeyKES says the caller gets to choose the hash, not the implementation. So that's why we have to hash the hash here. We could alternatively provide a "key identifier" function and let the implementation choose what that is.